h2/utilities.py

```diff
@@ -242,3 +249,56 @@ def _reject_pseudo_header_fields(headers):
             seen_regular_header = True
 
         yield header
+
+
+def _check_host_authority_header(headers, hdr_validation_flags):
+    """
+    Raises a ProtocolError if a header block arrives that does not contain
+    :authority or a Host header, or if a header block contains both fields,
+    but their values do not match.
+    """
+    # We only expect to see :authority and Host headers on request header
+    # blocks that aren't trailers, so skip this validation if we're on the
+    # server side or looking at trailer blocks.
+    if hdr_validation_flags.is_client or hdr_validation_flags.is_trailer:
+        for header in headers:
+            yield header
+        return
+
+    # We use None as a sentinel value.  Iterate over the list of headers,
+    # and record the value of these headers (if present).  We don't need
+    # to worry about receiving duplicate :authority headers, as this is
+    # enforced by the _reject_pseudo_header_fields() pipeline.
+    #
+    # @@AWLC: Do we have mechanisms to enforce non-duplicate Host headers?
+    authority_header_val = None
+    host_header_val = None
+
+    for header in headers:
+        if header[0] == b':authority':
+            authority_header_val = header[1]
+        elif header[0] == b'host':
+            host_header_val = header[1]
+
+        yield header
+
+    # If both of these variables are None, then we didn't find
+    # an :authority or a Host header.
+    if (
+        (authority_header_val is None) and
+        (host_header_val is None)
+    ):
+        raise ProtocolError(
+            "Did not receive an :authority or Host header."
+        )
+
+    # If we receive both headers, they should definitely match.
+    if (
+        (authority_header_val is not None) and
+        (host_header_val is not None) and:
+        (authority_header_val != host_header_val)
+    ):
```

Comment:
Sorry, I wasn't clear enough. I meant to have a separate `if` block. But now that I see both these conditionals together, I think there's a better way to structure this:

```python
authority_present = (authority_header_val is not None)
host_present = (host_header_val is not None)

if not authority_present and not host_present:
    raise ProtocolError(
        "Did not receive an :authority or Host header."
    )

if authority_present and host_present:
    if authority_header_val != host_header_val:
       raise ProtocolError("Mismatched")
```

This seems to me to most clearly communicate what's actually being tested. Do you agree with that assessment?